lessons/advanced/typespec.md

```diff
@@ -0,0 +1,127 @@
+---
+layout: page
+title: Specifications and types
+category: advanced
+order: 9
+lang: en
+---
+
+In this lesson we will learn about `@spec` and `@type` syntax. First is more syntax complement for writing documentation that could by analyzed by tools. Second helps us to write more readable and easier to understand code.
+
+{% include toc.html %}
+
+## Introduction 
+
+It's not uncommon you would like to describe interface of your function. Of course You can use [@doc annotation](/lessons/basic/documentation), but it is only information for other developers that is not checked in compilation time. For this purpose Elixir has `@spec` annotation to describe specification of function that will be checked by compiler.
+
+However in some cases specification is going to be quite big and complicated. If you would like to reduce complexity, you want to introduce custom type definition. Elixir has `@type` annotation for that. In the other hand, Elixir is still dynamic language. That means all information about type will be ignored by compiler, but could be used by other tools.   
+
+## Specification
+
+If you have experience with Java or Ruby you could think about specification as an `interface`. Specification defines what should be type of function parameters and return value.
+
+To define input and output types we use `@spec` directive placed right before function definition and taking as a `params` name of function, list of parameter types, and after `::` type of return value.  
+
+Let's take a look at example:
+
+```elixir
+@spec sum_product(integer) :: integer
+def sum_product(a) do
+    [1, 2, 3]
+    |> Enum.map(fn el -> el * a end)
+    |> Enum.sum
+end
+```
+
+Everything looks ok and when we call valid result will be return, but function `Enum.sum` returns `number` not `integer` as we expected in `@spec`. It could be source of bugs! There are tools like Dialyzer to static analysis of code that helps us to find this type of bugs. We will talk about them in another lesson.
+ 
+## Custom types
+
+Writing specification is nice, but sometimes our functions works with more complex data structures than simple numbers or collections. In that cases definition in `@spec` could be hard to understand and very hard to change by other developers. Sometimes functions needs to take many parameters or return complex data. Long parameters list is one of bad smells in code. In object oriented languages like Ruby or Java we could easily define classes that helps us to solve this problems. Elixir hasn't classes but because is easy to extends that we could define our types.
```

Comment:
Grammar/spelling suggestions:

`Writing specification`

-->

`Writing specifications`

================================

`In that cases definition in @spec could be hard to understand and very hard to change by other developers.`

-->

`In that definition's case in @spec it could be very hard to understand and/or change for other developers.`

================================

`Sometimes functions needs to take many parameters or return complex data.`

-->

`Sometimes functions need to take in a large number of parameters or return complex data.`

================================

`Long parameters list is one of bad smells in code.`

-->

`A long parameters list is one of many potential bad smells in one's code.`

================================

`In object oriented languages like Ruby or Java we could easily define classes that helps us to solve this problems.`

-->

`In object oriented-languages like Ruby or Java we could easily define classes that help us to solve this problem.`

================================

This one I just cant make sense of:

`Elixir has no concept of classes or object-orientation but because it is easy to extends that we could define our types.`